toolchain/check/member_access.cpp

```diff
@@ -303,26 +314,56 @@ static auto LookupMemberNameInScope(Context& context, SemIR::LocId loc_id,
-            witness_inst_id = GetOrAddInst(
-                context, loc_id,
-                SemIR::FacetAccessWitness{
-                    .type_id = GetSingletonType(
-                        context, SemIR::WitnessType::SingletonInstId),
-                    .facet_value_inst_id = base_id,
-                    .index = SemIR::ElementIndex(index)});
-            break;
+
+        bool is_lookup_in_period_self = false;
+        if (auto name = context.insts().TryGetAs<SemIR::NameRef>(base_id)) {
+          if (name->name_id == SemIR::NameId::PeriodSelf) {
+            is_lookup_in_period_self = true;
+          }
+        }
+
+        if (!is_lookup_in_period_self) {
+          // For an associated constant value, we need to do impl lookup to try
+          // find a final impl declaration. If we find one, we can use the value
+          // assigned to the constant there, instead of its symbolic value.
+          auto assoc_entity = context.insts().GetAs<SemIR::AssociatedEntity>(
+              context.constant_values().GetConstantInstId(
+                  result.scope_result.target_inst_id()));
+          if (context.insts().Is<SemIR::AssociatedConstantDecl>(
+                  assoc_entity.decl_id)) {
+            // TODO: These results should be cached somehow. Positive (non-None)
+            // results could be cached globally, as they can not change. But
+            // negative results can change after a final impl is written, so
+            // they can only be cached in a limited way, or the cache needs to
+            // be invalidated by writing a final impl that would match.
+            witness_inst_id = LookupFinalImplWitnessForSpecificInterface(
+                context, loc_id, context.constant_values().Get(base_id),
+                assoc_interface);
+          }
+        }
+
+        if (!witness_inst_id.has_value()) {
+          // First look for `assoc_interface` in the type of the base. If it is
+          // found, get the witness that the interface is implemented from
+          // `base_id`.
+          auto identified_id = RequireIdentifiedFacetType(context, *facet_type);
+          const auto& identified =
+              context.identified_facet_types().Get(identified_id);
+          for (auto [index, base_interface] :
+               llvm::enumerate(identified.required_interfaces())) {
+            // Get the witness that `T` implements `base_type_id`.
+            if (base_interface == assoc_interface) {
+              witness_inst_id = GetOrAddInst(
+                  context, loc_id,
+                  SemIR::FacetAccessWitness{
+                      .type_id = GetSingletonType(
+                          context, SemIR::WitnessType::SingletonInstId),
+                      .facet_value_inst_id = base_id,
+                      .index = SemIR::ElementIndex(index)});
+              break;
+            }
           }
         }
         // TODO: If that fails, would need to do impl lookup to see if the facet
```

Comment:
It seems like a big chunk of this handling could be replaced by a call to `LookupImplWitness`: we're doing a lookup for a final impl (which I think we need in `LookupImplWitness`) followed by a lookup within the facet type, followed by a TODO to perform the rest of impl lookup. Maybe an opportunity for a future refactoring?

Reply:
Yes, good point. I think.. this plays into the comment about `.Self` above as well.